Adding new tutorial to commerce-services (#347)
AdobeDocs/commerce-services: src/pages/shared-services/implementation/tutorial.md

                name: item.name,
                qty: item.qty,
                price: parseFloat(item.price)
            };
        });

        localStorage.setItem(STORAGE_KEY, JSON.stringify({ items: items }));
    }

    return function (originalAction) {
        return wrapper.wrap(originalAction, function (originalFn, paymentData, messageContainer) {
            persistCartData();
            return originalFn(paymentData, messageContainer);
        });
    };
});
```

## Step 6: Create the success-page component {#create-success-page}

Create `app/code/Vendor/CheckoutAnalytics/view/frontend/web/js/view/checkout-success.js`. This component sets [storefront context](../storefront-events/sdk/context.md) and restores cart data before it publishes the event.

```js
define([
    'magentoStorefrontEvents',
    'magentoStorefrontEventCollector'
], function () {
    'use strict';

    var STORAGE_KEY = 'mse_checkout_cart_data';

    return function (config) {
        var mse = window.magentoStorefrontEvents;
        var storedData = localStorage.getItem(STORAGE_KEY);
        var cartData = storedData ? JSON.parse(storedData) : null;

        if (!mse) {
            return;

>         }

Author: ignoring

The directory structure and code samples illustrate one implementation. Adjust module names, paths, and context values for your environment.

## Overview

You create a `Vendor_CheckoutAnalytics` module with these files:

| File | Purpose |
|---|---|
| `registration.php` | Registers the module with Adobe Commerce |
| `etc/module.xml` | Declares module dependencies |
| `view/frontend/requirejs-config.js` | Registers the mixin and SDK/collector CDN paths |
| `view/frontend/layout/checkout_onepage_success.xml` | Loads the success-page component |
| `view/frontend/templates/checkout-success.phtml` | Initializes the success-page component |
| `view/frontend/web/js/action/place-order-mixin.js` | Persists cart data before order placement |
| `view/frontend/web/js/view/checkout-success.js` | Restores context and publishes `placeOrder` |
| `view/frontend/web/js/noopSdk.js` | Fallback module when the SDK CDN fails to load |
| `view/frontend/web/js/noopCollector.js` | Fallback module when the collector CDN fails to load |
| `view/frontend/web/js/noopDs.js` | Fallback module when the data services CDN fails to load |

## Prerequisites

Before you begin, confirm the following:

> - Adobe Commerce 2.4.x with a Luma-based or Luma-derived frontend (RequireJS/AMD).

Reviewer: Do you need both RequireJS and AMD, or you use one or the other? (Based on the context of the tutorial, it looks like both are required)

```suggestion
- Adobe Commerce 2.4.x with a Luma-based or Luma-derived frontend (RequireJS and Asynchronous Module Definition (AMD))

```xml
<?xml version="1.0"?>
<!--
    Vendor_CheckoutAnalytics module declaration.
    Declares a soft dependency on Magento_Checkout so this module loads after it,
    ensuring our mixins target already-registered AMD modules.
-->
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:noNamespaceSchemaLocation="urn:magento:framework:Module/etc/module.xsd">
    <module name="Vendor_CheckoutAnalytics">
        <sequence>
            <module name="Magento_Checkout"/>
        </sequence>
    </module>
</config>
```

## Register the module

Create `app/code/Vendor/CheckoutAnalytics/registration.php` at the module root. This registers the module with the Magento component registry.

```php
<?php
/**
 * Vendor_CheckoutAnalytics
 *
 * Registers the module with the Magento component registry.
 */

use Magento\Framework\Component\ComponentRegistrar;

ComponentRegistrar::register(
    ComponentRegistrar::MODULE,
    'Vendor_CheckoutAnalytics',
    __DIR__
);
```

## Create fallback modules

> Create three empty AMD modules that serve as local fallbacks when the CDN is unreachable. RequireJS will automatically fall back to these if any CDN URL in `requirejs-config.js` fails to load, ensuring the checkout flow is never blocked.

Reviewer: ```suggestion
Create three empty AMD modules that serve as local fallbacks when the CDN is unreachable. RequireJS automatically falls back to these if any CDN URL in `requirejs-config.js` fails to load, ensuring the checkout flow is never blocked.
```

**File**: `app/code/Vendor/CheckoutAnalytics/view/frontend/web/js/noopSdk.js`:

```js
/**
 * Copyright &copy; Magento, Inc. All rights reserved.
 * See COPYING.txt for license details.
 */

/* eslint-disable */
define(function () { });
```

**File**: `app/code/Vendor/CheckoutAnalytics/view/frontend/web/js/noopCollector.js`:

```js
/**
 * Copyright &copy; Magento, Inc. All rights reserved.
 * See COPYING.txt for license details.
 */

/* eslint-disable */
define(function () { });
```

**File**: app/code/Vendor/CheckoutAnalytics/view/frontend/web/js/noopDs.js

```js
/**
 * Copyright &copy; Magento, Inc. All rights reserved.
 * See COPYING.txt for license details.
 */

/* eslint-disable */
define(function () { });
```

## Create the place-order mixin

> Create the place-order mixin. This intercepts `Magento_Checkout/js/action/place-order` before the redirect so cart and storefront data can be snapshotted to `localStorage`. Magento clears the cart server-side before the success page loads, making this the only opportunity to capture that data.

Reviewer: ```suggestion
Create the place-order mixin. This mixin intercepts `Magento_Checkout/js/action/place-order` before the redirect so cart and storefront data can be snapshotted to `localStorage`. Magento clears the cart server-side before the success page loads, making this the only opportunity to capture that data.
```

        const key = `CatalogView1:purchaseHistory`;
        const purchasedProducts = shoppingCartContext.items.map((item) => item.product.sku);
        const purchaseHistory = JSON.parse(window.localStorage.getItem(key) || '[]');
        purchaseHistory.push({ date: new Date().toISOString(), items: purchasedProducts });
        window.localStorage.setItem(key, JSON.stringify(purchaseHistory.slice(-20)));
        localStorage.removeItem(STORAGE_KEY);
    }

    return function () {
        console.log('[CheckoutAnalytics] checkout-success component initialized');

        var cartData = getPersistedCartData();
        if (!cartData) {
            console.warn('[CheckoutAnalytics] No cart snapshot in localStorage — ' +
                'placeOrder event will fire without item-level detail.');
        }

        var mse = window.magentoStorefrontEvents;
        if (!mse) {
            console.error('[CheckoutAnalytics] MSE SDK not available on window — cannot fire placeOrder event.');
            return;
        }

        console.log('[CheckoutAnalytics] cartData:', cartData);
        publishPlaceOrder(mse, cartData);
    };
});
```

## Register RequireJS configuration

> Create `requirejs-config.js`. This registers the place-order mixin and defines RequireJS paths for the MSE SDK, Collector, and data services base. Each path entry is an array — if the first CDN URL fails to load, RequireJS automatically falls back to the local noop module, so the checkout flow is never blocked.

Reviewer: ```suggestion
Create the `requirejs-config.js` configuration file. This registers the place-order mixin and defines RequireJS paths for the MSE SDK, Collector, and data services base. Each path entry is an array — if the first CDN URL fails to load, RequireJS automatically falls back to the local noop module, so the checkout flow is never blocked.
```

**File**: app/code/Vendor/CheckoutAnalytics/view/frontend/requirejs-config.js

```js
/**
 * RequireJS configuration for Vendor_CheckoutAnalytics.
 *
 * - place-order-mixin: snapshots cart to localStorage before the order redirect.
 * - paths: loads the MSE SDK and Collector as AMD deps with CDN + noop fallbacks.
 *   The success-page component (checkout-success.js) declares these as deps so

>  *   they are guaranteed to execute before its callback runs.

Reviewer: ```suggestion
 * - paths: loads the MSE SDK and Collector as AMD dependencies with CDN + noop fallbacks.
 *   The success-page component (checkout-success.js) declares these as dependencies so
 *   they are guaranteed to execute before its callback runs.
```

Deploy static content (the `-f` flag forces deployment in any application mode):

```bash
bin/magento setup:static-content:deploy -f
```

For production mode, specify your locale explicitly:

```bash
bin/magento setup:static-content:deploy en_US -f
```

Flush the cache:

```bash
bin/magento cache:flush
```

## Configure the storefront instance context

Open `view/frontend/web/js/view/checkout-success.js` and update the `setStorefrontInstance` call with your environment values:

```js
mse.context.setStorefrontInstance({
    environment:           'Production',   // 'Testing' or 'Production'
    environmentId:         'YOUR_ENV_ID',  // Tenant ID associated with ACO/EDS storefront
    baseCurrencyCode:      'USD',
    storeViewCurrencyCode: 'USD',
    viewId:                'YOUR_VIEW_ID', // Catalog View ID associated with ACO/EDS storefront
    storefrontTemplate:    'LUMA_BRIDGE',
    storeUrl:              'https://your-store.example.com'
});
```

<InlineAlert variant="important" slots="text"/>

> The `environmentId` value and `viewId` values need to align with the configuration of your EDS storefront. Using a placeholder value will cause events to be routed to the wrong data stream or dropped entirely.

Reviewer: ```suggestion
The `environmentId` value and `viewId` values need to align with the configuration of your EDS storefront. Using a placeholder value causes events to be routed to the wrong data stream or dropped entirely.
```

```

## Confirm module files

After you complete the steps above, confirm that your module directory contains the following files:

- `app/code/Vendor/CheckoutAnalytics/registration.php`
- `app/code/Vendor/CheckoutAnalytics/etc/module.xml`
- `app/code/Vendor/CheckoutAnalytics/view/frontend/requirejs-config.js`
- `app/code/Vendor/CheckoutAnalytics/view/frontend/layout/checkout_onepage_success.xml`
- `app/code/Vendor/CheckoutAnalytics/view/frontend/templates/checkout-success.phtml`
- `app/code/Vendor/CheckoutAnalytics/view/frontend/web/js/action/place-order-mixin.js`
- `app/code/Vendor/CheckoutAnalytics/view/frontend/web/js/view/checkout-success.js`
- `app/code/Vendor/CheckoutAnalytics/view/frontend/web/js/noopSdk.js`
- `app/code/Vendor/CheckoutAnalytics/view/frontend/web/js/noopCollector.js`
- `app/code/Vendor/CheckoutAnalytics/view/frontend/web/js/noopDs.js`

## Verify the implementation

### Cart snapshot on the checkout page

Confirm the mixin is capturing cart data before the redirect:

1. Open your store in a browser with the DevTools console open.
1. Add one or more products to the cart and proceed to checkout.

> 1. In the console, confirm the mixin initialized: `[CheckoutAnalytics] place-order mixin initialized`.

Reviewer: ```suggestion
1. In the console, confirm that the mixin initialized: `[CheckoutAnalytics] place-order mixin initialized`.
```

1. Place the order. Before the page redirects, confirm: `[CheckoutAnalytics] place-order mixin invoked, persisting cart data`.

> 1. In DevTools, open *Application > Local Storage* and confirm a key named `mse_checkout_cart_data` exists with a JSON payload containing cart items, totals, and storefront instance fields.

Reviewer: ```suggestion
1. In DevTools, open *Application > Local Storage* and confirm that a key named `mse_checkout_cart_data` exists with a JSON payload containing cart items, totals, and storefront instance fields.
```

### Event published on the success page

Confirm the `placeOrder` event fires correctly after the redirect:

> 1.    After the redirect to the success page, open the console and confirm the following log messages appear:

Reviewer: ```suggestion
1. After the redirect to the success page, open the console and confirm the following log messages appear:
```

```bash
[CheckoutAnalytics] checkout-success component initialized
[CheckoutAnalytics] cartData: { cartId: "...", items: [...], ... }
[CheckoutAnalytics] Publishing placeOrder event
```

> 1. Confirm the `mse_checkout_cart_data` key has been removed from Local Storage after the event fires.

Reviewer: ```suggestion
1. Confirm that the `mse_checkout_cart_data` key has been removed from Local Storage after the event fires.
```

> 1. To confirm the event was received by the MSE SDK, run the following in the console before placing a test order:

Reviewer: ```suggestion
1. To confirm that the event was received by the MSE SDK, run the following code in the console before placing a test order:
```

```bash
window.magentoStorefrontEvents.subscribe.placeOrder(function (event) {
    console.log('placeOrder event received:', event);
});
```

<InlineAlert variant="note" slots="text"/>

> If `window.magentoStorefrontEvents` is undefined on the success page, the CDN scripts failed to load. Check the **Network** tab in DevTools for failed requests to `cdn.jsdelivr.net`. The noop fallback modules prevent JavaScript errors but do not set `window.magentoStorefrontEvents`, so no event will fire.

Reviewer: ```suggestion
If `window.magentoStorefrontEvents` is undefined on the success page, the CDN scripts failed to load. Check the **Network** tab in DevTools for failed requests to `cdn.jsdelivr.net`. The noop fallback modules prevent JavaScript errors but do not set `window.magentoStorefrontEvents`, so no events fire.
```

## Troubleshooting

### The place-order mixin is not running

> Confirm the mixin is registered by checking the compiled RequireJS config in the browser:

Reviewer: ```suggestion
Confirm that the mixin is registered by checking the compiled RequireJS config in the browser:
```

```js
require.s.contexts._.config.config.mixins
// Should contain an entry for 'Magento_Checkout/js/action/place-order'
```

If the entry is missing, re-run static content deployment and flush the cache.

### mse_checkout_cart_data is empty or missing on the success page

> This means `persistCartData()` ran but the quote returned no items. This can happen if the active payment method bypasses the standard `place-order` action. Check whether your payment method uses a custom place-order action — if so, add a second mixin targeting that module.

Reviewer: ```suggestion
Missing or empty results mean that `persistCartData()` ran but the quote returned no items. This issue can happen if the active payment method bypasses the standard `place-order` action. Check whether your payment method uses a custom place-order action — if so, add a second mixin targeting that module.
```
